### PR TITLE
feat: Token management set up. Unify OAuth2 and access key.

### PR DESCRIPTION
**Pull request by shuofan**
#### What this PR does / why we need it:
Token management set up. Unify OAuth2 and access key.

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb25JRHMiOlsxMTc4LDExNzQsMTE5MF0sImFzc2lnbmVlIjpbIjEwMDEwNzMiXX0%3D&id=293897&iterationID=1174&pId=0&type=TASK)

#### Specified Reviewers:

/assign

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  Token management set up. Unify OAuth2 and access key.            |
| 🇨🇳 中文    |     token 管理         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: api/proto/core/token/token.proto [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      auth: {
+        check_login: true,
+        check_token: true,
+      }
+    };
+  }
+
+  rpc QueryTokens(QueryTokensRequest) returns (QueryTokensResponse) {
+    option (google.api.http) = {
+      get: "/api/tokens/actions/query",
```
Comment (Effet) [inline api/proto/core/token/token.proto:30]: why not simplify `/api/tokens` ?

File: conf/cluster-dialer/cluster-dialer.yaml [hunk 2]
```diff
@@ -9,6 +9,6 @@ http-server@admin:
     addr: ":9093"
 
 # core service credential
-grpc-client@erda.core.services.authentication.credentials.accesskey:
+grpc-client@erda.core.token:
   addr: "${CORE_SERVICES_GRPC_ADDR:core-services:9537}"
-erda.core.services.authentication.credentials.accesskey-client:
+erda.core.token-client:
```
Comment (Effet) [inline conf/cluster-dialer/cluster-dialer.yaml:14]: better: `erda.core.token-client: {}`

File: conf/cmp/cmp.yaml [hunk 3]
```diff
@@ -24,10 +24,9 @@ grpc-client@erda.core.monitor.alert:
 erda.core.monitor.alert-client:
 
 # core service credential
-grpc-client@erda.core.services.authentication.credentials.accesskey:
+grpc-client@erda.core.token:
   addr: "${CORE_SERVICES_GRPC_ADDR:core-services:9537}"
-  block: false
-erda.core.services.authentication.credentials.accesskey-client:
+erda.core.token-client:
```
Comment (Effet) [inline conf/cmp/cmp.yaml:29]: better: `erda.core.token-client: {}`

File: modules/cluster-dialer/auth/auth.go [hunk 4]
```diff
@@ -23,7 +23,7 @@ import (
 
 	"github.com/sirupsen/logrus"
 
-	credentialpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline modules/cluster-dialer/auth/auth.go:26]: alias `tokenpb` maybe better

File: modules/cluster-dialer/auth/auth.go [hunk 5]
```diff
@@ -88,11 +88,10 @@ func (a *Authorizer) Authorizer(req *http.Request) (string, bool, error) {
 	logrus.Debugf("get auth info: %s", authInfo)
 
 	// Doesn't need cache now
-	akSkResp, err := a.Credential.QueryAccessKeys(context.Background(), &credentialpb.QueryAccessKeysRequest{
-		AccessKey: authInfo,
-		Scope:     apistructs.CMPClusterScope,
-		ScopeId:   clusterKey,
-		Status:    credentialpb.StatusEnum_ACTIVATE,
```
Comment (Effet) [inline modules/cluster-dialer/auth/auth.go:95]: does `status` useless?

File: modules/cmp/impl/clusters/credential.go [hunk 6]
```diff
@@ -24,24 +24,24 @@ import (
...
 	"github.com/erda-project/erda/pkg/k8sclient"
+	"github.com/erda-project/erda/pkg/oauth2/tokenstore/mysqltokenstore"
 )
 
 // GetAccessKey get access key with cluster name.
-func (c *Clusters) GetAccessKey(clusterName string) (*credentialpb.QueryAccessKeysResponse, error) {
-	return c.credential.QueryAccessKeys(context.Background(), &credentialpb.QueryAccessKeysRequest{
-		Status:      credentialpb.StatusEnum_ACTIVATE,
-		SubjectType: credentialpb.SubjectTypeEnum_CLUSTER,
-		Subject:     clusterName,
```
Comment (Effet) [inline modules/cmp/impl/clusters/credential.go:39]: what difference between queryAccessKeys and QueryTokens, there Request look so different.

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go [hunk 7]
```diff
@@ -37,13 +40,61 @@ type TokenStore struct {
...
-	ExpiredAt *time.Time
-	Code      string
-	Access    string
-	Refresh   string
-	Data      TokenStoreItemData
+	ID            string `gorm:"primary_key"`
+	Scope         string
+	ScopeId       string
+	Access        string
+	SecretKey     string
```
Comment (Effet) [inline pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:47]: (`Access`, `SecretKey`), as the pair I noticed other system may call them `ID` and `Secret`?

Comment (Effet) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:47]: sometimes maybe they are duplicated?

**Review by Effet — CHANGES_REQUESTED** [review]

File: api/proto/core/token/token.proto (hunk 1 above)
Comment (shuofan) [reply api/proto/core/token/token.proto:30]: done

File: conf/cmp/cmp.yaml (hunk 3 above)
Comment (shuofan) [reply conf/cmp/cmp.yaml:29]: done

File: modules/cluster-dialer/auth/auth.go (hunk 4 above)
Comment (shuofan) [reply modules/cluster-dialer/auth/auth.go:26]: done

File: modules/cluster-dialer/auth/auth.go (hunk 5 above)
Comment (shuofan) [reply modules/cluster-dialer/auth/auth.go:95]: yes, it is useless now. I kept the property in db for future usage.

File: modules/cmp/impl/clusters/credential.go (hunk 6 above)
Comment (shuofan) [reply modules/cmp/impl/clusters/credential.go:39]: Cluster manage uses clusterName as ID. Subject and Subject type are deprecated. We use scope and scope id instead.

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go (hunk 7 above)
Comment (shuofan) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:47]: Yes, we use the "access key"(access key id) and "secret key"(secret access key) in AWS

File: conf/cluster-dialer/cluster-dialer.yaml (hunk 2 above)
Comment (shuofan) [reply conf/cluster-dialer/cluster-dialer.yaml]: done

File: modules/cmp/impl/clusters/credential.go (hunk 6 above)
Comment (Effet) [reply modules/cmp/impl/clusters/credential.go:39]: okay

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go (hunk 7 above)
Comment (Effet) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:47]: so I suggest to use `AccessKey` instead of `Access`?

Comment (Effet) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:47]: and `Scope` to `ScopeType` will more proper I think.

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go [hunk 8]
```diff
@@ -19,12 +19,15 @@ import (
...
 
+	"github.com/google/uuid"
 	"github.com/jinzhu/gorm"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/types/known/timestamppb"
 	"gopkg.in/oauth2.v3"
 	"gopkg.in/oauth2.v3/models"
 
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:30]: it's wired the `pkg/oauth2/tokenstore/mysqltokenstore` depend on `erda-proto-go`, maybe this pkg should not exist and all code move to `core-services` as plain CRUD.

Comment (Effet) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:30]: What do you think? @sfwn

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/cluster-dialer/auth/auth.go [hunk 9]
```diff
@@ -88,11 +88,10 @@ func (a *Authorizer) Authorizer(req *http.Request) (string, bool, error) {
...
+	akSkResp, err := a.Credential.QueryTokens(context.Background(), &tokenpb.QueryTokensRequest{
+		Access:  authInfo,
+		Scope:   apistructs.CMPClusterScope,
```
[the remark this answers is not in the document]
Comment (shuofan) [reply modules/cluster-dialer/auth/auth.go]: use enum in proto

File: api/proto/core/services/authentication/credentials/accesskey/accesskey.proto [hunk 10]
```diff
@@ -2,111 +2,7 @@ syntax = "proto3";
...
-  string            description = 3;
-}
-message UpdateAccessKeyResponse {}
-
-message DeleteAccessKeyRequest {
-  string id = 1;
-}
-message DeleteAccessKeyResponse {}
 
 message AccessKeysItem {
```
[the remark this answers is not in the document]
Comment (shuofan) [reply api/proto/core/services/authentication/credentials/accesskey/accesskey.proto]: Move to credential proto

File: modules/cluster-dialer/server/tunnel-server.go [hunk 11]
```diff
@@ -35,7 +35,7 @@ import (
 	"github.com/rancher/remotedialer"
 	"github.com/sirupsen/logrus"
 
-	credentialpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
[the remark this answers is not in the document]
Comment (shuofan) [reply modules/cluster-dialer/server/tunnel-server.go]: done

File: modules/cluster-dialer/provider.go [hunk 12]
```diff
@@ -21,14 +21,14 @@ import (
 	"github.com/sirupsen/logrus"
 
 	"github.com/erda-project/erda-infra/base/servicehub"
-	credentialpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
[the remark this answers is not in the document]
Comment (shuofan) [reply modules/cluster-dialer/provider.go]: done

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go (hunk 8 above)
Comment (shuofan) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:30]: Add TODO, will create a new pr for it.

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go (hunk 7 above)
Comment (shuofan) [reply pkg/oauth2/tokenstore/mysqltokenstore/token_store.go]: Changed to AccessKey. "Scope" includes definition of "type" when referred as the whole area which it deals with or includes.

File: modules/cmp/impl/clusters/credential.go [hunk 13]
```diff
@@ -17,31 +17,33 @@ package clusters
...
+	"strings"
 
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 
-	credentialpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline modules/cmp/impl/clusters/credential.go:28]: remove redundent import

File: modules/cmp/impl/clusters/upgrade_edge_cluster.go [hunk 14]
```diff
@@ -27,7 +27,7 @@ import (
 	"github.com/sirupsen/logrus"
 	"gopkg.in/yaml.v3"
 
-	credentialpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline modules/cmp/impl/clusters/upgrade_edge_cluster.go:30]: "tokenpb" is more proper

File: modules/cmp/initialize.go [hunk 15]
```diff
@@ -28,7 +28,7 @@ import (
 	"k8s.io/klog/v2"
 
 	"github.com/erda-project/erda-infra/base/version"
-	credentialpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline modules/cmp/initialize.go:31]: "tokenpb" is more proper

File: api/proto/core/token/token.proto [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+message GetTokenResponse {
+  Token data = 1;
+}
+
+message Token {
+  string id = 1;
+  string secretKey = 2;
+  string access = 3;
```
Comment (Effet) [inline api/proto/core/token/token.proto:92]: `accessKey` is more proper

File: pkg/oauth2/tokenstore/mysqltokenstore/token_store.go [hunk 17]
```diff
@@ -37,13 +42,61 @@ type TokenStore struct {
...
+		Type:        t.Type,
+		CreatorId:   t.CreatorID,
+		CreatedAt:   timestamppb.New(t.CreatedAt),
+	}
+}
+
+type TokenType string
+
+const (
+	Oauth2    TokenType = "Oauth2"
```
Comment (Effet) [inline pkg/oauth2/tokenstore/mysqltokenstore/token_store.go:90]: `OAuth2` is more proper

File: .erda/migrations/openapi/20220415-migrate-oauth2-token-base.sql [hunk 18]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO `erda_token` (id, code, access_key, refresh, data, type, created_at, updated_at, expired_at)
+SELECT UUID(), code, access, refresh, data, 'Oauth2', created_at, created_at, expired_at
```
Comment (Effet) [inline .erda/migrations/openapi/20220415-migrate-oauth2-token-base.sql:2]: `OAuth2` is more proper

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/oap/collector/authentication/validator.go [hunk 19]
```diff
@@ -18,10 +18,10 @@ import (
 	"context"
 	"sync"
 
-	akpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline modules/oap/collector/authentication/validator.go:21]: "tokenpb" is more proper

File: modules/oap/collector/authentication/provider.go [hunk 20]
```diff
@@ -20,7 +20,7 @@ import (
 
 	"github.com/erda-project/erda-infra/base/logs"
 	"github.com/erda-project/erda-infra/base/servicehub"
-	akpb "github.com/erda-project/erda-proto-go/core/services/authentication/credentials/accesskey/pb"
+	"github.com/erda-project/erda-proto-go/core/token/pb"
```
Comment (Effet) [inline modules/oap/collector/authentication/provider.go:23]: "tokenpb" is more proper

**Review by Effet — CHANGES_REQUESTED** [review]

**Review by Effet — APPROVED** [review]

**Pull request merged**